cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/relational/SQLDialectConverter.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright © 2023 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.datapipeline.relational;
+
+import io.cdap.cdap.api.data.schema.Schema;
+
+import javax.annotation.Nullable;
+
+/**
+ * An SPI that declares a service for conversion of a SQL expression from one SQL dialect to another.
+ * The {@link SQLDialectConversionExpressionFactory} uses this service to perform the conversion.
+ */
+public interface SQLDialectConverter {
+
+    /**
+     * Validates the specified SQL expression using the schema given for the specified SQL dialect
+     *
+     * @param expression A SQL expression in the form of a full statement in the SQL dialect specified
+     * @param srcDialect The SQL dialect in which the SQL statement is specified
+     * @param schema The schema for the table being used in the SQL expression
+     * @throws SQLDialectException in case the SQL expression is invalid, or any other error related to the SQL
+     * dialect validation is encountered
+     */
+    void validate(String expression, SQLDialect srcDialect, @Nullable Schema schema) throws SQLDialectException;
```

Comment:
I noticed that we always fail if the schema is null. Should we change the contract then and ensure that the schema is always non-null for this class?

Reply:
This is mostly a consideration for if / when a second non-Calcite implementation is added which might not need a schema for validation or conversion (although thinking of it I can't quite see how validation might be possible without a schema).

Reply:
I'm inclined to think that the schema **must** be known to be able to validate that the supplied expression is valid. For example, column names and data types must be known to make sure that the expression is valid and correct.

Reply:
I agree. Removed the `@Nullable` annotation.